### DISCUSSION TITLE
Nginx Snippets don't support websockets

### DISCUSSION THREAD

**Discussion by jamesshannon** [authelia/authelia#5350] [Q&A] [answered]
### Version

v4.37.5

### Deployment Method

Docker

### Reverse Proxy

NGINX

### Description

The [documented nginx snippets](https://github.com/authelia/authelia/blob/master/docs/content/en/integration/proxies/nginx.md#supporting-configuration-snippets) breaks wss in nginx for me. I ran into this on my own install when trying to proxy teslamate. The wss connection did not work but did work when I removed the authelia configuration.

I'm using Nginx Proxy Manager, but I don't think that's relevant. The [solution on the web](https://www.reddit.com/r/selfhosted/comments/nxfymg/authelia_nginx_proxy_manager_issues_with_vscode/) is to add:

```
 proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection "upgrade";
```

to the configuration. 

I notice that the `proxy.conf` snippet actually overrides the `Connection` header. 

Thus I believe that including the two `proxy_set_header` lines above in the snippet would solve this for others. The reason I'm opening this issue and being verbose is because I don't understand enough about authelia, nginx, or wss to be sure that this doesn't break something, or if there is a reason that the existing snippet sets the Connection header to an empty string. However, adding those two lines to the `proxy.conf` file worked for me.

I'll also note that nginx has ["A more sophisticated example in which a value of the “Connection” header field in a request to the proxied server depends on the presence of the “Upgrade” field in the client request header"](http://nginx.org/en/docs/http/websocket.html) in case that might be more appropriate.

### Reproduction

1. Set up authelia + nginx + [service that uses wss]
2. Authenticate and load page

### Logs (Authelia)

```shell
na
```

### Pre-Submission Checklist

- [X] I agree to follow the [Code of Conduct](http://www.authelia.com/code-of-conduct)
- [X] This is a bug report and not a support request
- [X] I have read the security policy and this bug report is not a security issue or security related issue
- [X] I have either included the complete configuration file or I am sure it's unrelated to the configuration
- [X] I have provided all of the required information in full with the only alteration being reasonable sanitization in accordance with the [Troubleshooting Sanitization](https://www.authelia.com/r/sanitize) reference guide
- [X] I have checked for related proxy or application logs and included them if available
- [X] I have checked for related issues and checked the documentation

**Comment by james-d-elliott** [comment]
Setting the desired connection header after the include would resolve this issue if it were an issue related to this line, additionally provided it doesn't break the general use case the most appropriate adjustment is to remove this from our proxy.conf example and users should configure this on a per-app basis as per their individual obligation to adapt the configuration to their needs.

**Comment by james-d-elliott** [comment]
Are you able to confirm removal of the header in question fixes this?

**Comment by tylercollier** [comment]
Thanks to @jamesshannon for providing the solution.

I removed the directive originally from the `proxy.conf` example:

```
proxy_set_header Connection "";
```

And, using the article he linked (using the "more sophisticated example"), replaced it with these lines in `proxy.conf`:

```
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection $connection_upgrade;
```

**Comment by remisharrock** [comment]
I confirm the 

```
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection "upgrade";
```

works for me too.

@tylercollier the $connection_upgrade  variable was unrecognized by my nginx

**Comment by tylercollier** [comment]
@remisharrock In the [article he linked](https://nginx.org/en/docs/http/websocket.html), it has this snippet:

```nginx
map $http_upgrade $connection_upgrade {
    default upgrade;
    ''      close;
}
```

**Comment by tylercollier** [comment]
To future readers (as in, me): For what it's worth, make sure you don't list these directives twice:

```nginx
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection "upgrade";
```

When I accidentally did (I had them in my site config as well as the snippet), web sockets did not work.

Oddly, nginx warns about certain duplicate directives. Perhaps this one is different because it's headers, and technically some headers can appear multiple times. I'm not sure if the actual problem was due to nginx or Node (that was my true backend) though.

**Comment by svenwanzenried** [comment]
EDIT: See see the conversation below. This first solution is unnecessary complicated...

I want to add my solution when WSS was not working with Nginx Proxy Manager, Authelia and Portainer (shell exec in container would not connect).
Due to a quirk in NPM or nginx, you have to add an entry in "Custom locations" whenever you want to set new headers in a location. You **can not** just use the "Advanced" tab.
So my working config looks like this:
![image](https://github.com/user-attachments/assets/ef7a5544-e80e-4516-bbae-b12797426eeb)
![image](https://github.com/user-attachments/assets/b52132b3-e268-4c9d-94ee-98703e7d6c43)
![image](https://github.com/user-attachments/assets/f0d636f4-47be-40b6-8d38-133c70dd43ae)

which will produce an NGINX config file like this:

```nginx
# ------------------------------------------------------------
# portainer.domain.tld
# ------------------------------------------------------------
map $scheme $hsts_header {
    https   "max-age=63072000;includeSubDomains; preload";
}

server {
  set $forward_scheme https;
  set $server         "portainer";
  set $port           9443;
  listen 80;
  listen [::]:80;
  listen 443 ssl http2;
  listen [::]:443 ssl http2;
  server_name portainer.domain.tld;
  
  # Let's Encrypt SSL
  include conf.d/include/letsencrypt-acme-challenge.conf;
  include conf.d/include/ssl-ciphers.conf;
  ssl_certificate /etc/letsencrypt/live/npm-1/fullchain.pem;
  ssl_certificate_key /etc/letsencrypt/live/npm-1/privkey.pem;

  # Block Exploits
  include conf.d/include/block-exploits.conf;
  # HSTS (ngx_http_headers_module is required) (63072000 seconds = 2 years)
  add_header Strict-Transport-Security $hsts_header always;

  # Force SSL
  include conf.d/include/force-ssl.conf;
  proxy_set_header Upgrade $http_upgrade;
  proxy_set_header Connection $http_connection;
  proxy_http_version 1.1;
  access_log /data/logs/proxy-host-4_access.log proxy;
  error_log /data/logs/proxy-host-4_error.log warn;

  include /snippets/authelia-location.conf;

  location / {
    location / {
    include /snippets/proxy.conf;
    include /snippets/authelia-authrequest.conf;
    proxy_pass $forward_scheme://$server:$port;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
  }

  proxy_set_header Host $host;
  proxy_set_header X-Forwarded-Scheme $scheme;
  proxy_set_header X-Forwarded-Proto  $scheme;
  proxy_set_header X-Forwarded-For    $remote_addr;
  proxy_set_header X-Real-IP          $remote_addr;
  set $proxy_forward_scheme https;
  set $proxy_server         "portainer";
  set $proxy_port           9443;
  proxy_pass       $proxy_forward_scheme://$proxy_server:$proxy_port;
    
    
  # Access Rules: 1 total
  allow all;
  deny all;

  # Access checks must...
  satisfy all;

  # Block Exploits
  include conf.d/include/block-exploits.conf;
    
  # Force SSL
  include conf.d/include/force-ssl.conf;
    
  # HSTS (ngx_http_headers_module is required) (63072000 seconds = 2 years)
  add_header Strict-Transport-Security $hsts_header always;
    
  proxy_set_header Upgrade $http_upgrade;
  proxy_set_header Connection $http_connection;
  proxy_http_version 1.1;
    
  }
  # Custom
  include /data/nginx/custom/server_proxy[.]conf;
}
```

**Reply by james-d-elliott** [reply]
Thanks for this. Can you confirm that everything operates correctly with both websocket hosts and non-websocket hosts? Also be good if you know what a PWA is and if this is compatible. It'd be good to include this in the base configs if it's not going to break anything for anyone but it's been difficult to get people to confirm this information.

**Reply by svenwanzenried** [reply]
after a good nights sleep, i can not confirm anything it seems.. When i looked at my config again, it appeared to me that there is a nested "location /" clause. I don't know the exact implications of that, but I guess it's not 'proper' at least.
I have to make more experiments as how to configure this using a combination of advanced tab and custom location.

regarding PWA, I currently do not host one, so I cant test this right away. Do you know any selfhosting friendly service that is a PWA and can be installed easily?

**Reply by james-d-elliott** [reply]
Try changing the custom location to (i.e. no location block, and in your second screenshot):

```nginx
include /snippets/proxy.conf;
include /snippets/authelia-authrequest.conf;
proxy_pass $forwarded_scheme://$server:$port;
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection "upgrade";
```

I am fairly `lldap` is a PWA but I'm not absolutely certain, it is easy to setup though.

**Reply by svenwanzenried** [reply]
I tried that already.. Also various different versions, but it turns out it needs this nested location to work with Portainer at all.
I think, it's because there must not be any repeating statements in the same location. And NPM keeps inserting some headers after the custom location.
That leads to errors in NGINX. So with this second nested clause, the headers set by NPM (in the first "location /" clause) get ignored and it works.

Currently i have no idea, how to circumvent that without going so deep into NPM config files that I would probably be easier to just use NGINX without the NPM manager.

**Reply by james-d-elliott** [reply]
I would have thought if you just did it in the advanced section similar to how the guide shows but adding the additional headers it'd be happy. Pretty sure they document if you make adjustments to the root location in the advanced section you get what you add.

**Reply by svenwanzenried** [reply]
Thanks for beeing persistent 👍
I turns out, that it works just with the advanced section. I don't really know, why yesterday it wasn't working, because I am sure, I tried it this way. Probably something with the browser cache.

This configuration in the advanced section works now for Portainer (even tested in Private Browsing to avoid cache issues):
```nginx
  include /snippets/authelia-location.conf;

  location / {
    include /snippets/proxy.conf;
    include /snippets/authelia-authrequest.conf;
    proxy_pass $forward_scheme://$server:$port;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    #proxy_http_version 1.1; # this must be disabled if "Websockets Support" is enabled on the first dialog tab. Otherwise it's redundant and nginx complains!
  }
```

If i have time, I'll give LLDAP a look concerning PWA...

**Comment by svenwanzenried** [comment]
@james-d-elliott I just tried lldap. But i could not find that it's a PWA. So I setup a miniflux container. This is a Feed-Reader that has a PWA. I could start it and i configured a proxy host in Nginx Proxy Manager (NPM) with the Authelia NGINX snippets like so:
![image](https://github.com/user-attachments/assets/0e06ba65-4e64-4fae-8712-3ed967acd6da)
![image](https://github.com/user-attachments/assets/2f724025-fc76-4b5c-b6e6-ede52f4e8bf6)

I could perform the following actions successfully:
- Reach login page
- Login
- Download the PWA
- Add feeds in the PWA
- Read feed articles in the PWA
- Log out from Authelia in the browser
- Click a link in the PWA, which redirected me to the Authelia login-page IN the PWA
- Login successfully with 2FA to Authelia in the PWA, which redirected me back to the PWA

So in conclusion, this seems to work also in an PWA. However, what I don't know is if miniflux uses websockets. I doubt it, because if I disabled WS support in NPM, the Feed reader still continues to work. I hope this helps 😊
If you point me to a PWA thats easy to setup (docker) and uses Websockets, then I would be happy to help testing again.

**Reply by james-d-elliott** [reply]
Nice, I'll make the adjustments when I get time, thanks for your responsiveness.

**Reply by james-d-elliott** [reply, marked as answer]
I've added a [PR](https://github.com/authelia/authelia/pull/8496) with a websocket example. It should be noted that as per [nginx support](https://trac.nginx.org/nginx/ticket/2115) the Connection/Upgrade headers should not be sent on all requests and we've adapted our examples accordingly.
